cli/diagnostics.rs

```diff
@@ -0,0 +1,640 @@
+    let start_position = obj.get("startPosition").and_then(Value::as_i64);
+    let end_position = obj.get("endPosition").and_then(Value::as_i64);
+    let start_column = obj.get("startColumn").and_then(Value::as_i64);
+    let end_column = obj.get("endColumn").and_then(Value::as_i64);
+
+    let message_chain_v = obj.get("messageChain");
+    let message_chain = match message_chain_v {
+      Some(v) => DiagnosticMessageChain::from_json_value(v),
+      _ => None,
+    };
+
+    let related_information_v = obj.get("relatedInformation");
+    let related_information = match related_information_v {
+      Some(r) => {
+        let mut related_information = Vec::<DiagnosticItem>::new();
+        let related_info_values = r.as_array().unwrap();
+
+        for related_info_v in related_info_values {
+          related_information
+            .push(DiagnosticItem::from_json_value(related_info_v));
+        }
+
+        Some(related_information)
+      }
+      _ => None,
+    };
+
+    Self {
+      message,
+      message_chain,
+      related_information,
+      code,
+      source_line,
+      script_resource_name,
+      line_number,
+      start_position,
+      end_position,
+      category,
+      start_column,
+      end_column,
+    }
+  }
+}
+
+impl DisplayFormatter for DiagnosticItem {
+  fn format_category_and_code(&self) -> String {
+    let category = match self.category {
+      DiagnosticCategory::Error => {
+        format!("- {}", ansi::red("error".to_string()))
+      }
+      DiagnosticCategory::Warning => "- warn".to_string(),
+      DiagnosticCategory::Debug => "- debug".to_string(),
+      DiagnosticCategory::Info => "- info".to_string(),
+      _ => "".to_string(),
+    };
+
+    let code = ansi::grey(format!(" TS{}:", self.code.to_string())).to_string();
+
+    format!("{}{} ", category, code)
+  }
+
+  fn format_message(&self, level: usize) -> String {
+    if self.message_chain.is_none() {
+      return format!("{:indent$}{}", "", self.message, indent = level);
+    }
+
+    let mut s = String::new();
+    let mut i = level / 2;
+    let mut item_o = self.message_chain.clone();
+    while item_o.is_some() {
+      let item = item_o.unwrap();
+      s.push_str(&std::iter::repeat(" ").take(i * 2).collect::<String>());
+      s.push_str(&item.message);
+      s.push('\n');
+      item_o = item.next.clone();
+      i += 1;
+    }
+    s.pop();
+
+    s
+  }
+
+  fn format_related_info(&self) -> String {
+    if self.related_information.is_none() {
+      return "".to_string();
+    }
+
+    let mut s = String::new();
+    let related_information = self.related_information.clone().unwrap();
+    for related_diagnostic in related_information {
+      let rd = &related_diagnostic;
+      s.push_str(&format!(
+        "\n{}{}{}\n",
+        rd.format_source_name(2),
+        rd.format_source_line(4),
+        rd.format_message(4),
+      ));
+    }
+
+    s
+  }
+
+  fn format_source_line(&self, level: usize) -> String {
```

Comment:
Add a TODO to share formatting implementation with cli/js_errors.rs ... there's not exactly an equivalent function in that module but `js_errors::format_script_line_column()` should be applicable